# Exploring the gene_exp.diff data from cuffdiff with stats turned on

```{r message = FALSE}
library(tidyverse)
library(data.table)
library(ggfortify)
library(ggplot2)
library(factoextra)
if (!require(ComplexHeatmap)) { BiocManager::install("ComplexHeatmap") }
library(ComplexHeatmap)
library(pvclust)
library(parallel)
library(dendextend)
```


```{r read in dataset}
#setwd("~/Desktop/TRGN514/rnaseq_proj/")

exp <- data.table::fread("./data/gene_exp.diff", sep = "\t")
```

```{r}
head(exp)
tail(exp)
dim(exp) #62635 tests (number of genes tested)
```

```{r filter dataset}
# the cutoff will be p-val of < 0.001 and log2 fold change greater than 5 or less than -5

exp_filt <- exp %>%
  arrange(`log2(fold_change)`) %>%
  filter(p_value < 0.001) %>%
  filter(`log2(fold_change)` != -Inf, # Inf or -Inf means gene had 0 expression in one of the groups
         `log2(fold_change)` != Inf) %>%
  filter(`log2(fold_change)` <= -5 | `log2(fold_change)` >= 5)
```


```{r check data}
nrow(exp_filt)
head(exp_filt)
```

```{r}
up_genes <- exp_filt %>%
  filter(`log2(fold_change)` > 0) %>%
  select(gene_id, gene)

down_genes <- exp_filt %>%
  filter(`log2(fold_change)` < 0) %>%
  select(gene_id, gene)
```


FPKM

```{r}
fpkm <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm)[26:29] <- names(fpkm)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm <- fpkm %>% 
  select(tracking_id, ends_with("_FPKM"))
```


```{r pca preprocessing}
row.names(fpkm) <- fpkm$tracking_id

fpkm <- fpkm %>% select(ends_with("FPKM"))

fpkm <- t(fpkm)

# normalize everything using log conversion
fpkm_norm <- fpkm + 0.1

fpkm_norm <- log10(fpkm_norm)

# create sample column for labeling
fpkm_norm <- as.data.frame(fpkm_norm)
fpkm_norm$samples <- row.names(fpkm_norm)

fpkm_norm.data <- fpkm_norm[-ncol(fpkm_norm)]
```

```{r}
fpkm_norm$type <- ifelse(grepl("norm", fpkm_norm$samples), "normal tissue",
                            ifelse(grepl("ct", fpkm_norm$samples), "cancer tissue", "cancer cell line"))

```


```{r plot pca edit to scale}
pca <- prcomp(fpkm_norm.data, scale. = FALSE)

pca$x

autoplot(pca, scale = FALSE, data = fpkm_norm, colour = "type") + ggtitle("PCA unfiltered")
```



Plot PCA
```{r plot pca no edits to scale, eval=FALSE}
autoplot(prcomp(fpkm_norm.data), data = fpkm_norm, colour = "type")
```


Boxplots for all gene FPKM values across all samples

```{r fpkm boxplots of all samples across all genes}
df <- t(fpkm) %>%
  as.data.frame()

df <- df[rowSums(df)>0,] #remove genes with 0 expression across all samples

df <- df + 0.1


group.colors <- c(norm1_FPKM = "#2D95F6", norm2_FPKM = "#2D95F6", 
                  ct1_FPKM = "#FFFB14", ct2_FPKM = "#FFFB14", ct3_FPKM = "#FFFB14", 
                  ccl1_FPKM = "#FE2323", ccl2_FPKM = "#FE2323", ccl3_FPKM = "#FE2323")

reshape2::melt(df) %>%
  ggplot(aes(x = variable, y = log10(value), fill = variable)) +
  geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_fill_manual(values = group.colors)
```

PCA with filtered genes only

```{r redo PCA with filtered genes only}
fpkm_filt <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm_filt)[26:29] <- names(fpkm_filt)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm_filt <- fpkm_filt %>% 
  select(tracking_id, ends_with("_FPKM")) %>%
  filter(tracking_id %in% exp_filt$test_id) %>%
  column_to_rownames("tracking_id")


fpkm_filt_norm <- log10(fpkm_filt + 0.1)          

fpkm_filt_norm <- t(fpkm_filt_norm)

fpkm_filt_norm <- fpkm_filt_norm %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  mutate(type = ifelse(grepl("norm", sample), "normal tissue",
                         ifelse(grepl("ct", sample), "cancer tissue", "cancer cell line")))
  

fpkm_filt_norm.data <- fpkm_filt_norm %>% select(-c(sample, type))

pca2 <- prcomp(fpkm_filt_norm.data)

pca2$x

autoplot(pca2, scale = FALSE, data = fpkm_filt_norm, colour = "type") + ggtitle("PCA on Filtered Genes")
```

Compute Dendrograms

Use pearson correlation for distance and ward.D2 and complete for clustering method

```{r Dendrogram of differentially expressed genes ward method, warning = FALSE}
rownames(fpkm_filt_norm.data) <- fpkm_filt_norm$sample

res.pearson.dist <- get_dist(fpkm_filt_norm.data, method = "pearson")
res.euclidean.dist <- get_dist(fpkm_norm.data, method = "euclidean")

res.ward.hc <- hclust(d = res.euclidean.dist, method = "ward.D2")
fviz_dend(res.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Method:", "ward.D2", sep = " "))


res.complete.hc <- hclust(d = res.euclidean.dist, method = "complete")
fviz_dend(res.complete.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Method:", "complete", sep = " "))
```



Evaluate Dendrograms

```{r p-values for ward clustering}
# create cluster for parallel computing
n.cores <- detectCores()*3/4
cl <- makePSOCKcluster(n.cores)
clusterSetRNGStream(cl, 1234)

# calculate p-value dendrogram for ward.d2 method
res.ward.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "ward.D2", method.dist = "correlation", nboot = 1000)

# plot
plot(res.ward.pv, hang = -1, cex = 0.5)

# Clusters with AU >= 95% are considered to be strongly supported by data.
```

```{r p-values for complete clustering}
n.cores <- detectCores()*3/4
cl <- makePSOCKcluster(n.cores)
clusterSetRNGStream(cl, 1234)

# calculate p-value dendrogram for complete method
res.complete.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "complete", method.dist = "correlation", nboot = 1000)
stopCluster(cl)

# plot
plot(res.complete.pv, hang = -1, cex = 0.5)

```

```{r}
dend1 <- as.dendrogram(res.complete.hc)
dend2 <- as.dendrogram(res.ward.hc)

dendlist(dend1, dend2) %>%
  untangle(method = "step1side") %>%
  tanglegram(main_left = "complete", main_right = "ward.D2")
```




```{r heatmap complete}
Heatmap(fpkm_filt_norm.data, 
        show_column_names = FALSE,
        row_split = 3)
```


```{r heatmap wards}
Heatmap(fpkm_filt_norm.data, 
        name = "expression",
        show_column_names = FALSE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")
```






