---
title: "GI Cancer Gene Expression Data Visualization"
author: "Flemming Wu"
output: github_document
always_allow_html: true
---

## Introduction

In this project, I used publicly available RNA-Seq data from the Gene Expression Omnibus database (accession number GSE41476) to look at how differential gene expression between three groups of samples: gastric cancer tissue, matched adjacent normal tissue, and gastric cancer cell lines. 

The process of downloading raw sequencing data, alignment, and obtaining gene expression counts was all done on a HPC server and the commands used for that can be found in the "server_commands.sh" file on this repository. This file describes my downstream analysis steps.

-----

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.height=6, fig.width=8, fig.align='center')
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
rm(list=ls())
```

```{r load libraries, message = FALSE, warning=FALSE}
library(tidyverse)
library(data.table)
library(ggfortify)
library(ggplot2)
library(factoextra)
library(ComplexHeatmap)
library(pvclust)
library(parallel)
library(dendextend)
library(scales)
library(DiagrammeR)
library(gridExtra)
library(tidytext)
library(readxl)
library(circlize)
library(webshot)
```

------

#### Plot Read Counts For all 8 Samples

The tissue samples I worked with in my project come from the bio-resource center of the Asan Medical Center in Seoul. The researchers collected human gastric cancer samples and adjacent normal tissues. They also collected gastric cancer cell lines from the Korean Cell Line Bank.


There were a total of 8 samples between the three groups: 2 normal gastric tissue samples, 3 gastric cancer tissue samples, and 3 samples from gastric cancer cell lines. The samples were all paired-end, and sequenced on Illumina’s Genome Analyzer II and stored in SRA format.

```{r pre-alignment metric plot, fig.width=10}
samples <- c("SRR585570", "SRR585571", "SRR585572", "SRR585573", "SRR585574", "SRR585575", "SRR585576", "SRR585577")

reads <- c(18883105, 16580585, 27917182, 34687136, 32772474, 34049244, 31929290, 32457417)

groups <- c("normal gastric tissue", "normal gastric tissue", "gastric cancer tissue", "gastric cancer tissue", "gastric cancer tissue", "gastric cancer cell line", "gastric cancer cell line", "gastric cancer cell line")

df <- data.frame(samples, reads, groups)

df %>%
  ggplot(aes(y = reads/1000000, x = samples)) + 
  geom_col(aes(fill = groups)) +
  scale_fill_manual(values=c("#23E7F7",
                             "#2336F7",
                             "#EEF51F")) + 
  geom_text(aes(label = format(reads, big.mark = ",")), position = position_dodge(width = 1), vjust = -0.5) +
  geom_hline(yintercept = mean(reads)/1000000, color = "black", linetype = "dotted") +
  geom_text(aes(0, mean(reads)/1000000 , label = paste("Average reads:", format(round(mean(reads)), big.mark = ","), sep = " "), vjust = -1, hjust = 0)) +
  labs(
       x = "Sample",
       y = "Number of Reads (millions)",
       fill = "Group"
       ) 
```

I determined that I was able to use all 8 samples, since they all had a good number of reads for my purposes. The minimum number of reads was 16.5 million which was in the second normal tissue sample, the maximum was 34.6 million which was in my second gastric cancer tissue sample, and all 8 samples averaged at 28.6 million reads.

------

#### Flowchart of bioinformatic workflow

```{r Flowchart of bioinformatic workflow}
grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Arial, shape = rectangle, color = Lavender, style = filled]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
      tab8 [label = '@@8']
      tab9 [label = '@@9']

      # edge definitions with the node IDs
      tab1 -> tab3;
      tab2 -> tab3;
      tab3 -> tab4; 
      tab3 -> tab5;
      tab4 -> tab6;
      tab4 -> tab7;
      tab6 -> tab8;
      tab7 -> tab9;
      }

      [1]: 'Download read (FASTQ) files from SRA using SRA-Toolkit'
      [2]: 'Download reference genome (FASTA) and annotation (GTF) files from Ensembl'
      [3]: 'Build index with Bowtie2'
      [4]: 'Align reads to whole genome using Tophat2 (BAM)'
      [5]: 'Align reads to whole transcriptome using Tophat2 (BAM)'
      [6]: 'Obtain FPKM values using Cuffdiff with statistics turned off'
      [7]: 'Obtain p-values and log2 fold differences between normal and cancer tissue using Cuffdiff with statistics on'
      [8]: 'Output file: genes.fpkm_tracking'
      [9]: 'Output file: gene_exp.diff'
      ")
```

The above commands were all performed on a HPC cluster. The output files, genes.fpkm_tracking and gene_exp.diff were the files I used for downstream analysis in R and are provided in the data folder in this repository. 

I first downloaded the SRA files as fastq files using the fasterq-dump command from sra-toolkit. I also downloaded the fasta file for the human reference genome, version GRCh38, and the GTF file of the genome annotation, both of which were from Ensembl. 

I then built index files from the reference genome using Bowtie2, which will be used to speed up the mapping process. Next, I selected one normal tissue sample and one cancer tissue sample to map to the whole genome and the whole transcriptome using Tophat2 to see if they mapped to one much better. It turned out that the mapping percentages for both genome and transcriptome were both at around 80%, so I just mapped the rest of my samples to the whole genome. 

Once all the reads were mapped, I took my alignment files, or bam files, and put them into two different runs of Cuffdiff. In the first run, I inputted all 8 of my samples and ran Cuffdiff with statistics turned off to get just the read counts of all the genes. The output for this was in my genes.fpkm_tracking file. In the second run, I gave Cuffdiff only my normal and cancer tissue samples as two groups and ran it with the statistics on to get log2 fold change and p-values for the genes just between these two groups. The output for this was in the gene_exp.diff file.

-----

#### Post-Tophat2 Alignment Metrics

```{r alignment metrics, fig.width=9}
df <- df %>%
  mutate(input = c(37766210, 33161170, 55834364, 69374272, 65544948, 68098488, 63858580, 64914834),
         mapped = c(31498668, 27402515, 45797812, 58072751, 52477113, 55003238, 42375461, 44831741),
         pct_mapped = round(mapped/input * 100, 1)
         )

df %>%
  ggplot(aes(y = pct_mapped, x = samples)) + 
  geom_col(aes(fill = groups)) +
  scale_fill_manual(values=c("#23E7F7", "#2336F7", "#EEF51F")) +
  geom_text(aes(label = paste(pct_mapped, "%", sep = " ")), position = position_dodge(width = 1), vjust = -0.5) +
  labs(x = "Sample",
       y = "Percent of Reads Aligned to Genome",
       fill = "Group")
```

Overall, the Tophat alignment was pretty successful. Most of my samples have over an 80% mapping rate. The maximum mapping percentage is 83.7%, and the two samples on the right in the chart are cancer cell lines 2 and 3, they had slightly lower mapping rates than the other samples, at about 66 and 69%. But, the average across all 8 samples was 78.5%. 

-----

#### PCA analysis on all genes across all samples

```{r read in and process data}
fpkm <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm)[26:29] <- names(fpkm)[26:29] %>% # fix column names that were incorrectly named in Cuffdiff command
  str_replace("ct2", "ct3")


# select needed columns

fpkm <- fpkm %>% 
  select(tracking_id, ends_with("_FPKM"))

row.names(fpkm) <- fpkm$tracking_id

fpkm <- fpkm %>% select(ends_with("FPKM"))

fpkm <- t(fpkm)


# normalize everything using log10 conversion

fpkm_norm <- fpkm + 0.1

fpkm_norm <- log10(fpkm_norm)


# create sample column for labeling

fpkm_norm <- as.data.frame(fpkm_norm)
fpkm_norm$samples <- row.names(fpkm_norm)

fpkm_norm.data <- fpkm_norm[-ncol(fpkm_norm)]

fpkm_norm$type <- ifelse(grepl("norm", fpkm_norm$samples), "normal tissue",
                            ifelse(grepl("ct", fpkm_norm$samples), "cancer tissue", "cancer cell line"))

```


```{r plot pca and edit to scale}
pca <- prcomp(fpkm_norm.data, scale. = FALSE)

#pca$x # view principal components 

autoplot(pca, scale = FALSE, data = fpkm_norm, colour = "type", size = 3) + 
  ggtitle("PCA on 40,295 Genes") + 
  theme_bw()
```

You can see that the groups clustered together pretty nicely already, with the cancer tissue on the right, the normal tissue on the left, and the cancer cell lines on the bottom. PC1 accounts for 31.2% of the variance in the data set and seems to separate cancer from non-cancer groups. PC2 accounts for 24.8% of the variance in the data and seems to separate tissues and cell lines.

-----

#### Boxplots for all gene FPKM values across all samples

```{r prepare data for boxplot}
df <- as.data.frame(fpkm)
df <- df[colSums(df) > 0] # filter out genes with 0 expression across all samples
df <- df + 0.1
df <- log10(df)
```

```{r fpkm boxplots of all samples across all genes, warning=FALSE}
df %>%
  t() %>%
  reshape::melt() %>%
  arrange(X1) %>%
  rename(sample = X2) %>%
  mutate(group = ifelse(grepl("norm", sample), "normal tissue", ifelse(grepl("ct", sample), "cancer tissue", "cancer cell line"))) %>%
  ggplot(aes(x = sample, y = value, fill = group)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Sample", y = "log10(FPKM)", fill = "Group", title = "FPKM Distributions for 40,295 Genes") +
  theme_bw() 
```

From the plot, it looks like the highest median fpkm is in the normal tissue group by just a little bit, but the overall distribution of gene expression looks very similar across all 8 samples. 

-----

My next step was to filter genes down based on the log2 fold change and p-values so that I could have a manageable number of genes to look at for pathway analysis. The cutoffs I ended up using were a p-value of less than 0.001 and log 2 fold change that was less than -5 or greater than 5. The filtering left me with 280 significantly differentially expressed genes, 198 of them were downregulated and 82 of them were upregulated.

#### Diagram on filtering criteria for Integrated Pathway Analysis (from QIAGEN)

```{r gene filtering}
grViz(diagram = "digraph flowchart {
      # define node aesthetics
      node [fontname = Arial, shape = egg, color = DarkGreen, style = filled, fontcolor = White]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
# set up node layout
      tab1 -> tab6;
      tab6 -> tab7;
      tab7 -> tab2;
      tab2 -> tab3;
      tab3 -> tab4;
      tab3 -> tab5
      }
      [1]: 'Initial number of genes: 62,635'
      [2]: 'Filter p-value < 0.001 and log2 fold change < -5 or > 5'
      [3]: '280 genes Remaining'
      [4]: '198 genes downregulated'
      [5]: '82 genes upregulated'
      [6]: 'Filter 0 gene expression across all 8 samples'
      [7]: '40,295 genes remaining'
      ")
```

 

Perform filtering of genes according to the above diagram

```{r read in expression differences file and filter}

exp <- data.table::fread("./data/gene_exp.diff", sep = "\t")

# the cutoff will be p-val of < 0.001 and log2 fold change greater than 5 or less than -5

exp_filt <- exp %>%
  arrange(`log2(fold_change)`) %>%
  filter(p_value < 0.001) %>%
  filter(`log2(fold_change)` != -Inf, # Inf or -Inf means gene had 0 expression in one of the groups
         `log2(fold_change)` != Inf) %>%
  filter(`log2(fold_change)` <= -5 | `log2(fold_change)` >= 5)
```

Separate up and down regulated genes

```{r separate upregulated and downregulated genes}
up_genes <- exp_filt %>%
  filter(`log2(fold_change)` > 0) %>%
  select(gene_id, gene)

down_genes <- exp_filt %>%
  filter(`log2(fold_change)` < 0) %>%
  select(gene_id, gene)
```

-----

#### PCA with filtered genes only

```{r redo PCA with filtered genes only}
fpkm_filt <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm_filt)[26:29] <- names(fpkm_filt)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm_filt <- fpkm_filt %>% 
  select(tracking_id, ends_with("_FPKM")) %>%
  filter(tracking_id %in% exp_filt$test_id) %>%
  column_to_rownames("tracking_id")


fpkm_filt_norm <- log10(fpkm_filt + 0.1)          

fpkm_filt_norm <- t(fpkm_filt_norm) %>% as.data.frame()

fpkm_filt_norm$sample <- row.names(fpkm_norm)

fpkm_filt_norm <- fpkm_filt_norm %>% 
  mutate(type = ifelse(grepl("norm", sample), "normal tissue",
                         ifelse(grepl("ct", sample), "cancer tissue", "cancer cell line")))


fpkm_filt_norm.data <- fpkm_filt_norm %>% select(-c(sample, type))

pca2 <- prcomp(fpkm_filt_norm.data, scale. = FALSE)

#pca2$x

autoplot(pca2, scale = FALSE, data = fpkm_filt_norm, colour = "type", size = 3) + 
  ggtitle("PCA on 280 Filtered Genes") + 
  theme_bw()
```

With the 280 filtered genes, I redid my principal component analysis and got the plot that you see on the screen. It looks very similar to my first PCA plot with the unfiltered genes, the three groups also clustered together and in the same areas, but in this one it looks like the samples within each group clustered a bit closer together. Also, principal component 1 now accounts for 62% of the variance in the data, which is double the principal component 1 from my original PCA plot.

-----

#### Compute distance measures and clustering methods for hierarchical clustering

```{r Dendrogram of differentially expressed genes ward method, warning = FALSE}
rownames(fpkm_filt_norm.data) <- fpkm_filt_norm$sample

# try pearson correlation and euclidean distance for distance measures
res.pearson.dist <- get_dist(fpkm_filt_norm.data, method = "pearson")
res.euclidean.dist <- get_dist(fpkm_norm.data, method = "euclidean")

# try ward and complete hierarchical clustering methods
res.eucl.ward.hc <- hclust(d = res.euclidean.dist, method = "ward.D2")
res.eucl.complete.hc <- hclust(d = res.euclidean.dist, method = "complete")
res.pear.ward.hc <- hclust(d = res.pearson.dist, method = "ward.D2")
res.pear.complete.hc <- hclust(d = res.pearson.dist, method = "complete")


p1 <- fviz_dend(res.pear.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "pearson", "Method:", "ward.D2", sep = " "))


p2 <- fviz_dend(res.pear.complete.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "pearson", "Method:", "complete", sep = " "))

p3 <- fviz_dend(res.eucl.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "euclidean", "Method:", "ward.D2", sep = " "))

p4 <- fviz_dend(res.eucl.complete.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "euclidean", "Method:", "complete", sep = " "))

grid.arrange(p1, p2, p3, p4)
```

It appears that using euclidean distance along with Ward's method gives us the clustering that we expect to see given our knowledge of the three groups.

```{r visualize p3, warning=FALSE}
fviz_dend(res.eucl.ward.hc, cex = 0.75, k = 3, color_labels_by_k = TRUE, k_colors = c("blue", "red", "darkgreen"),
          rect = TRUE, rect_fill = TRUE, show_labels = TRUE,
          main = "Hierarchical Clustering of Samples Based on 280 Filtered Genes", xlab = "Samples") 
```

The top separation separated cancer from non-cancer, and within the cancer cluster there's a separation between cancer tissue and cancer cell lines.



#### Evaluate Dendrograms using p-values 

```{r p-values for ward clustering, warning=FALSE, message=FALSE}
# create cluster for parallel computing
n.cores <- detectCores()*3/4
cl <- makePSOCKcluster(n.cores)
clusterSetRNGStream(cl, 1234)

# calculate p-value dendrogram for pearson correlation and ward.d2 method
res.pear.ward.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "ward.D2", method.dist = "correlation", nboot = 1000)
# calculate p-value dendrogram for pearson correlation and complete method
res.pear.complete.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "complete", method.dist = "correlation", nboot = 1000)
# calculate p-value dendrogram for euclidean dist and ward.d2 method
res.eucl.complete.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "ward.D2", method.dist = "euclidean", nboot = 1000)

stopCluster(cl)

# plot
## Clusters with AU >= 95% are considered to be strongly supported by data.
plot(res.pear.ward.pv, hang = -1, cex = 0.5)
plot(res.pear.complete.pv, hang = -1, cex = 0.5)
plot(res.eucl.complete.pv, hang = -1, cex = 0.5)
```

Above, it appears that using pearson correlation and complete clustering method clustered ccl3_FPKM poorly. The AU values for using both pearson and euclidean along with Ward's method gave stronger AU values.


#### Compare top 2 dendrograms using tanglegram

```{r view tanglegram of top clusters}
dend1 <- as.dendrogram(res.pear.ward.hc)
dend2 <- as.dendrogram(res.eucl.ward.hc)

dendlist(dend1, dend2) %>%
  untangle(method = "step1side") %>%
  tanglegram(main_left = "Pearson Correlation", main_right = "Euclidean Distance", 
             main = "Ward.D2", lwd = 1, cex_main = 1, margin_inner = 5)
```

Using pearson correlation for distance metric instead of euclidean results in cancer cell line 2 clustering with the normal tissues instead of with the other cancer cell line samples. I have determined that using euclidean distance and ward's method to be the best choices with my data.


-----

#### Plot heatmap of gene expression

```{r annotated heatmap wards}
boxplot <- rowAnnotation(`log10(fpkm)` = anno_boxplot(data.matrix(fpkm_filt_norm.data)),
                         gp=gpar(fill="#CCCCCC"))

hmap <- Heatmap(data.matrix(fpkm_filt_norm.data), 
        name = "Expression",
        show_column_names = FALSE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2",
        right_annotation = boxplot)

draw(hmap)
```

You can see that there is a large block of genes that are expressed highly in normal tissue and a block of genes expressed highly in cancer tissue. The cell line cluster seems to be spread out, with some highly expressed genes overlapping with normal tissue and some overlapping with cancerous tissue. On the right side, I also included boxplots of the overall fpkm values and you can see that the expression levels are higher overall for the normal samples, compared to the cancer samples which makes sense because now I’m working with more genes that are downregulated in cancer than upregulated


-----

At this point, I looked to the list of upregulated and downregulated genes from tthe integrated pathway analysis (software from QIAGEN) to find a pathway to look at more in depth. There were many significant pathways returned, many of the down regulated ones were involved in immune system function whereas many of the up regulated pathways have been previously implicated in cancer. (The excel files are located in the IPA-results folder.)

Moving forward, I will be look at the pathway title "metabolism of carbohydrates", which was downregulated in cancerous samples, compared to normal samples.



#### Plot heatmap of carbohydrate metabolism pathway.

```{r get list of genes in metabolism of carbohydrates pathway}
meta <- read_xls("./IPA-results/down-genes_diseases_and_functions.xls", col_names = TRUE) %>%
  filter(`Diseases or Functions Annotation` == "Metabolism of carbohydrate") %>%
  unnest_tokens(genes, Molecules) %>%
  mutate(genes = toupper(genes)) %>%
  select(genes)

meta %>% head(22)
```

```{r create df containing FPKM values for genes in carbohydrate metabolism pathway}
fpkm <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm)[26:29] <- names(fpkm)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm <- fpkm %>% 
  select(gene_short_name, ends_with("_FPKM"))

meta <- merge(x = meta,
      y = fpkm, 
      all.x = TRUE,
      all.y = FALSE,
      by.x = "genes",
      by.y = "gene_short_name")


meta <- meta %>% 
  column_to_rownames("genes") %>% 
  t()


meta_norm <- log10(meta + 0.1)
```


```{r annotated heatmap, warning=FALSE}
Heatmap(meta_norm, 
        name = "expression",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2") + 
  rowAnnotation(`log10(fpkm)` = anno_boxplot(meta_norm)) +
  rowAnnotation(rn = anno_text(rownames(meta_norm)))
```



Using the same clustering method (euclidean distance and ward.D2 method) it appears that cancer cell line 2 becomes its own cluster, and cancer cell lines 1 and 3 cluster with the cancer tissue.

```{r other clustering method testing}
# compare other clustering methods with original to check to see if it is still the best method for the selected pathway

# original clustering method
ht1 <- Heatmap(meta_norm, 
        name = "ht1",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")


# pearson correlation with ward.D2 clustering
ht2 <- Heatmap(meta_norm, 
        name = "ht2",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2",
        column_names_gp = gpar(fontsize = 9))

# euclidean distance with complete clustering 

ht3 <- Heatmap(meta_norm, 
        name = "ht3",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "euclidean",
        clustering_distance_columns = "euclidean",
        clustering_method_rows = "complete",
        clustering_method_columns = "complete",
        column_names_gp = gpar(fontsize = 9))

# pearson correlation with ward.D2 clustering

ht4 <- Heatmap(meta_norm, 
        name = "ht4",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_rows = "complete",
        clustering_method_columns = "complete",
        column_names_gp = gpar(fontsize = 9))

ht1 + ht2 + ht3 + ht4
```

When plotting all 4 heatmaps together with the original heatmap as the main, the separation is clearest with the original. Individually, ht2 makes cell line 3 its own cluster and clusters cell line 2 and cancer tissue 3 with normal, ht2 is similar to original heatmap, and ht3 makes cell line 1 and 3 are their own clusters. I conclude that the original clustering method is still the best option for this pathway.

-----

#### Linear model test on individual genes for significance

```{r add principal components to data frame}

# add PC1 and PC2 to the data frame for correction

lm_df <- merge(x = pca$x, y = meta_norm, by = "row.names") %>%
  select(-c(4:9)) %>%
  rename(group = Row.names) %>%
  mutate(group = if_else(grepl("norm", group), "normal tissue", if_else(grepl("ct", group), "cancer tissue", "cancer cell line")))

#lm_df
```

```{r lm test}
# function to return p-value from lm object
lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}


# lm test on all genes
no_correction <- list() # initialize empty list to hold p-values for non-corrected lm test
correction <- list() # empty list to hold p-values for corrected lm test
for (i in names(lm_df[4:length(names(lm_df))])){
  no_correction[i] <- lmp(lm(get(i) ~ group, lm_df))
  correction[i] <- lmp(lm(get(i) ~ group + PC1 + PC2, lm_df))
}


# store results in data frame
meta_pval <- rbind(as.data.frame(no_correction), as.data.frame(correction)) %>% 
  t() %>%
  as.data.frame() %>%
  rename("Non PC Corrected p-value" = V1, "PC Corrected p-value" = V2) %>%
  format(scientific = FALSE) %>%
  rownames_to_column("gene")

meta_pval <- meta_pval %>%
  mutate(`Non PC Corrected p-value` = as.numeric(`Non PC Corrected p-value`),
         `PC Corrected p-value` = as.numeric(`PC Corrected p-value`)) 

```


Plot p-values for each gene's significance in determining difference between cancer and non cancer


```{r plot p-values, message=FALSE}
meta_pval %>% 
  arrange(`PC Corrected p-value`) %>%
  rename(`p-value before PC correction` = `Non PC Corrected p-value`,
         `p-value after PC correction` = `PC Corrected p-value`) %>%
  reshape::melt() %>%
  ggplot(aes(x = fct_inorder(gene), y = value)) +
  geom_bar(aes(fill = variable), stat = "identity", position = "dodge", width = 0.5) +
  scale_fill_manual(values = c("gray", "#FA871B")) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_hline(yintercept = 0.05, linetype = "dashed") +
  labs(y = "p-value", x = "gene") 
```

It appears that after correcting for the first two principal components, all genes except for CD22, B3GNT8, B3GAT1, and ALDH1A1 are significant.


-----

#### Quick information about significant genes:

* P2RY6: [promotes tumorigenesis by inhibiting apoptosis](https://pubmed.ncbi.nlm.nih.gov/29454075/)

* PIK3CG: [inhibits PI3K-Akt/PKB signaling system responsible for tumorigenesis and the progression of colorectal cancers](https://pubmed.ncbi.nlm.nih.gov/12473596/)

* **GPD1**: [encodes glycerol-3-phosphate dehydrogenase, which plays a crucial role in carbohydrate and lipid metabolism by catalyzing conversion of DHAP and NADG to G3P (glycerol-3-phosphate) and NAD+.](https://en.wikipedia.org/wiki/Glycerol-3-phosphate_dehydrogenase_1) Has been implicated as a tumor suppressor in [breast cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5731876/) and [bladder cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9284842/)

* CSFR1: [involved in production, differentiation, and function of macrophages](https://en.wikipedia.org/wiki/Colony_stimulating_factor_1_receptor)

* **GCNT4**: [predicted to be involved in O-glycan processing and carbohydrate metabolic process.](https://www.ncbi.nlm.nih.gov/gene/51301). [Found to be downregulated in gastric cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7457769/)

* PRKCB: [encodes protein involved in B cell activation, apoptosis induction, endothelial cell proliferation, and intestinal sugar absorption.](https://www.ncbi.nlm.nih.gov/gene/5579) [Previously implicated in colorectal cancer](https://link.springer.com/article/10.1007/s12253-013-9739-5)

* B4GALNT2: [catalyzes last step in biosynthesis of human Sd(a) antigen and Cad antigen.](https://www.genecards.org/cgi-bin/carddisp.pl?gene=B4GALNT2) [Found to be highly expressed in model triple negative breast cancer](https://pubmed.ncbi.nlm.nih.gov/34589428/)

* CEBPA: [encodes transcription factor involved in maturation of certain blood cells and believed to act as a tumor supressor.](https://medlineplus.gov/genetics/gene/cebpa/) [Implicated in acute myeloid leukemia](https://www.frontiersin.org/articles/10.3389/fonc.2022.806137/full)

* SST: [regulator of endocrine system through interactions with pituitary growth hormone, thyroid stimulating hormone, and most hormones of GI tract. Also affects rates of neurotransmission in the CNS and proliferation of normal and tumorigenic cells](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SST)

* **SLC2A4**: [encodes member of K+ dependent Na/Ca exchanger protein family](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SLC24A4) Not many articles on its implications in cancer.

* PTPRC: [known to be signaling molecules that regulate a variety of cellular processes including cell growth, differentiation, mitosis, and oncogenic transformation](https://www.ncbi.nlm.nih.gov/gene/5788)

* GCNT2: [encodes enzyme responsible for formation of blood group I antigen.](https://en.wikipedia.org/wiki/GCNT2) [Hypomethylated in colorectal cancer](https://pubmed.ncbi.nlm.nih.gov/25750292/)

-----

I have selected three genes of interest to me: GPD1, GCNT4, and SLC24A for further analysis.


```{r barplots of individual genes, fig.height=12, fig.width=12}
p1 <- lm_df %>%
  select(group, GPD1) %>%
  group_by(group) %>%
  summarise(mean = mean(GPD1), sd = sd(GPD1)) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(group), mean), y = mean, fill = group)) +
  geom_col() +
  geom_errorbar(aes(x = group, ymin = mean - sd, ymax = mean + sd), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_text(aes(label = paste("Average log10(fpkm):", round(mean, 2), sep = " "),
                hjust = 0.5,
                vjust = ifelse(mean < 0, 5.5, -2.5)),
            size = 3) +
  geom_hline(yintercept = 0) +
  ylim(-2, 2) +
  labs(x = "", y = "Average log(10) FPKM", title = "GPD1") +
  annotate("text", label = paste("p-value:", round(lmp(lm(GPD1 ~ group, lm_df)), 6), sep = " "), x = 2, y = 2)

p2 <- lm_df %>%
  select(group, SLC2A4) %>%
  group_by(group) %>%
  summarise(mean = mean(SLC2A4), sd = sd(SLC2A4)) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(group), mean), y = mean, fill = group)) +
  geom_col() +
  geom_errorbar(aes(x = group, ymin = mean - sd, ymax = mean + sd), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_text(aes(label = paste("Average log10(fpkm):", round(mean, 2), sep = " "),
                hjust = 0.5,
                vjust = ifelse(mean < 0, 5.5, -2.5)),
            size = 3) +
  geom_hline(yintercept = 0) +
  ylim(-2, 2) +
  labs(x = "", y = "Average log(10) FPKM", title = "SLC2A4") +
  annotate("text", label = paste("p-value:", round(lmp(lm(SLC2A4 ~ group, lm_df)), 6), sep = " "), x = 2, y = 2)


p3 <- lm_df %>%
  select(group, GCNT4 ) %>%
  group_by(group) %>%
  summarise(mean = mean(GCNT4 ), sd = sd(GCNT4 )) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(group), mean), y = mean, fill = group)) +
  geom_col() +
  geom_errorbar(aes(x = group, ymin = mean - sd, ymax = mean + sd), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_text(aes(label = paste("Average log10(fpkm):", round(mean, 2), sep = " "),
                hjust = 0.5,
                vjust = ifelse(mean < 0, 5.5, -2.5)),
            size = 3) +
  geom_hline(yintercept = 0) +
  ylim(-2, 2) +
  labs(x = "", y = "Average log(10) FPKM", title = "GCNT4 ") +
  annotate("text", label = paste("p-value:", round(lmp(lm(GCNT4 ~ group, lm_df)), 6), sep = " "), x = 2, y = 2)

grid.arrange(p1, p2, p3, ncol = 2)
```

Here are the barplots of their average expression and standard error between the three groups. I also included the p-values of their significance after correcting for principal components 1 and 2, you can see that levels of expression are a lot lower in cancer cell lines and tissue compared to the normal groups and their error bars don’t overlap.

-----

#### SLC2A4

SLC2A4 is a member of the solute-carrier family-2 proteins. It encodes the protein GLUT4, which is responsible for the diffusion of circulating glucose down its concentration gradient into muscle and fat cells, where they are converted either to glucose-6-phosphate for glycolysis or glycogen for storage. 

GLUT4 plays an essential role in maintaining body glucose homeostasis and is regulated by insulin

In a [previous study](https://pubmed.ncbi.nlm.nih.gov/34488531/) looking at the solute-carrier family-2’s implications in cancer, they analyzed mRNA expression data in 20 cancer types from the Oncomine database and filtered based on a p-value of less than 0.01 and fold change of 1.5. They found that SLC2A4 was significantly downregulated in breast cancer. 

To verify their results, they used a web-based tool called UALCAN and found that mRNA expression of SLC2A4 was significantly downregulated in all individual breast cancer stages with p-value < 0.01. I included a screenshot of their results in the top picture.

They then used a Kaplan-Meier survival analysis and demonstrated that high expression of SLC2A4 was related to better prognosis in stage I and III breast cancer. I included a picture of their survival curve for stage I cancer on the bottom.

Lastly, they used a software called MEXPRESS to assess whether DNA hypermethylation in the promoter of SLC2A4 was related to its expression. They found that there was a statistical difference when comparing breast cancer samples versus normal samples, suggesting that this is a regulator of mRNA expression. 

Overall, they suggest that GLUT4 downregulation may inhibit glucose uptake and induce metabolic reprogramming and also affect downstream pathways that play key roles in tumorigenesis and progression. But they state that further studies with larger sample sizes are needed for a better understanding of the mechanisms of SLC2A4 in cancer.


#### GPD1

GPD1 plays a role in carbohydrate and lipid metabolism, specifically in the conversion of DHAP and NADH into glycerol-3-phosphate and NAD+. Together with GPD2, the two proteins carry electrons generated by glycolysis into the electron transport chain. 

In [this study](https://jhoonline.biomedcentral.com/articles/10.1186/s13045-022-01312-5) that used 17 pairs of bladder cancer tissue and matched adjacent tissue, they analyzed differential expression using proteomics, and identified almost 100,000 peptides. After filtering with 1.5 fold change difference and a p-value of 0.05 as the cutoff, they narrowed the proteins down to 518 downregulated proteins, and found GPD1 to be the most significantly downregulated, which I highlighted in the volcano plot in figure B. 

They verified their results using a Western blot (figure C on the slide) and also used a Kaplan-Meier estimate to observe that GPD1 was correlated with decreased survival time for breast cancer patients, which led them to hypothesize that GPD1 may have antitumor functions.

To summarize their findings: they determined that GPD1 increases a compound called lysoPC through a glycerophospholipid metabolism pathway. lysoPC then upregulated expression of TRPV2, causing a calcium influx that inhibits cell proliferation and induces apoptosis.



#### GCNT4

GCNT 4 stands for glucosaminyl (N-acetyl) transferase 4. It is part of a family, named GCNT1 through 4, and all of them are involved in mucin core structure synthesis and branching. These mucins are the most common macromolecules in mucus and are mainly responsible for its properties.

It’s been hypothesized that members of the GCNT family influences cancer genesis by regulating cell growth and apoptosis pathways and have been identified in colon, breast, prostate, and pancreatic cancer.

In [this paper](https://pubmed.ncbi.nlm.nih.gov/34696660/), the researchers looked at gene expression data from gastric cancer and non-tumor samples published in GEO. They filtered differentially expressed genes using the criteria of having a log fold change of less than -1.5 and p-value of 0.05 and found GCNT4 as a gene of interest as a suppressor of gastric cancer cell proliferation. 

Additionally, they used a program called TargetScan to predict an microRNA that binds to the 3’UTR of GCNT4 and confirmed in 45 of their own gastric cancer samples that this microRNA decreases GCNT4 expression which in turn upregulated downstream pathways that have been studied in the progression of gastric cancer.


-----

### Conclusions and Future Directions

Overall, my findings of abnormally expressed genes and pathways in gastric cancer were consistent with my expectations with many cancer related pathways being upregulated and many immune cell pathways being downregulated. 

Also, the downregulation of carbohydrate metabolism pathways has been observed in a variety of different cancers, including gastric cancer, and the three genes that I looked at are involved in different parts of carbohydrate metabolism pathways. 

It's possible that the pathway I looked at was significant in gastric cancer because of the genes’ “cross-talk” with other pathways that have to do with cell cycle or immune cell regulation.

Moving forward, I think that it would be insightful to screen a larger and more diverse sample of gastric cancer tissues and include people of different ethnicities as well as from the different subtypes of gastric cancer so that we can see if we observe any overlap in carbohydrate metabolism pathways. Then, I think that it would be interesting to filter out the pathways that overlap with ones that are abnormally expressed in other types of cancer, so that we can consider any that are left to be possibly gastric cancer-specific.

