---
title: "GI Cancer Gene Expression Data Visualization"
author: "Flemming Wu"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.height=6, fig.width=8, fig.align='center')
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
rm(list=ls())
```

```{r load libraries, message = FALSE, warning=FALSE}
library(tidyverse)
library(data.table)
library(ggfortify)
library(ggplot2)
library(factoextra)
library(ComplexHeatmap)
library(pvclust)
library(parallel)
library(dendextend)
library(scales)
library(DiagrammeR)
library(gridExtra)
library(tidytext)
library(readxl)
library(circlize)
```

#### Plot Read Counts For all 8 Samples

```{r pre-alignment metric plot, fig.width=10}
samples <- c("SRR585570", "SRR585571", "SRR585572", "SRR585573", "SRR585574", "SRR585575", "SRR585576", "SRR585577")

reads <- c(18883105, 16580585, 27917182, 34687136, 32772474, 34049244, 31929290, 32457417)

groups <- c("normal gastric tissue", "normal gastric tissue", "gastric cancer tissue", "gastric cancer tissue", "gastric cancer tissue", "gastric cancer cell line", "gastric cancer cell line", "gastric cancer cell line")

df <- data.frame(samples, reads, groups)
#df

df %>%
  ggplot(aes(y = reads, x = samples)) + 
  geom_col(aes(fill = groups)) +
  scale_fill_manual(values=c("#23E7F7",
                             "#2336F7",
                             "#EEF51F")) +
  scale_y_continuous(labels = scales::comma) +
  geom_hline(yintercept = mean(reads), color = "red", linetype = "dotted") + 
  geom_text(aes(label = format(reads, big.mark = ",")), position = position_dodge(width = 1), vjust = -0.5) +
  labs(caption = paste(paste("Minimum reads:", format(min(reads), big.mark = ","), sep = " "),
                       paste("Average reads:", format(mean(reads), big.mark = ","), sep = " "),
                       paste("Maximum reads:", format(max(reads), big.mark = ","), sep = " "),
                       sep = "   "),
       x = "Sample",
       y = "Number of Reads",
       fill = "Group"
  ) +
  theme(plot.caption.position = "plot")

```

#### Draw flowchart describing general bioinformatic workflow

```{r Flowchart of bioinformatic workflow}
grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Arial, shape = rectangle, color = Lavender, style = filled]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
      tab8 [label = '@@8']
      tab9 [label = '@@9']

      # edge definitions with the node IDs
      tab1 -> tab3;
      tab2 -> tab3;
      tab3 -> tab4; 
      tab3 -> tab5;
      tab4 -> tab6;
      tab4 -> tab7;
      tab6 -> tab8;
      tab7 -> tab9;
      }

      [1]: 'Download read (FASTQ) files from SRA using SRA-Toolkit'
      [2]: 'Download reference genome (FASTA) and annotation (GTF) files from Ensembl'
      [3]: 'Build index with Bowtie2'
      [4]: 'Align reads to whole genome using Tophat2 (BAM)'
      [5]: 'Align reads to whole transcriptome using Tophat2 (BAM)'
      [6]: 'Obtain FPKM values using Cuffdiff with statistics turned off'
      [7]: 'Obtain p-values and log2 fold differences between normal and cancer tissue using Cuffdiff with statistics on'
      [8]: 'Output file: genes.fpkm_tracking'
      [9]: 'Output file: gene_exp.diff'
      ")
```

#### Post-Tophat2 Alignment Metrics

```{r alignment metrics, fig.width=9}
df <- df %>%
  mutate(input = c(37766210, 33161170, 55834364, 69374272, 65544948, 68098488, 63858580, 64914834),
         mapped = c(31498668, 27402515, 45797812, 58072751, 52477113, 55003238, 42375461, 44831741),
         pct_mapped = round(mapped/input * 100, 1)
         )

df %>%
  ggplot(aes(y = pct_mapped, x = samples)) + 
  geom_col(aes(fill = groups)) +
  scale_fill_manual(values=c("#23E7F7", "#2336F7", "#EEF51F")) +
  geom_text(aes(label = paste(pct_mapped, "%", sep = " ")), position = position_dodge(width = 1), vjust = -0.5) +
  labs(x = "Sample",
       y = "Percent of Reads Aligned to Genome",
       fill = "Group")
```


#### PCA analysis on all genes across all samples

```{r read in and process data}
fpkm <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm)[26:29] <- names(fpkm)[26:29] %>% # fix column names that were incorrectly named in Cuffdiff command
  str_replace("ct2", "ct3")


fpkm <- fpkm %>% 
  select(tracking_id, ends_with("_FPKM"))

row.names(fpkm) <- fpkm$tracking_id

fpkm <- fpkm %>% select(ends_with("FPKM"))

fpkm <- t(fpkm)

# normalize everything using log10 conversion
fpkm_norm <- fpkm + 0.1

fpkm_norm <- log10(fpkm_norm)

# create sample column for labeling
fpkm_norm <- as.data.frame(fpkm_norm)
fpkm_norm$samples <- row.names(fpkm_norm)

fpkm_norm.data <- fpkm_norm[-ncol(fpkm_norm)]

fpkm_norm$type <- ifelse(grepl("norm", fpkm_norm$samples), "normal tissue",
                            ifelse(grepl("ct", fpkm_norm$samples), "cancer tissue", "cancer cell line"))

```


```{r plot pca and edit to scale, fig.width=7}
pca <- prcomp(fpkm_norm.data, scale. = FALSE)

pca$x

autoplot(pca, scale = FALSE, data = fpkm_norm, colour = "type") + ggtitle("PCA") + theme_bw()
```

#### Boxplots for all gene FPKM values across all samples

```{r prepare data for boxplot}
df <- as.data.frame(fpkm)
df <- df[colSums(df) > 0] # filter out genes with 0 expression across all samples
df <- df + 0.1
df <- log10(df)
```

```{r fpkm boxplots of all samples across all genes}
df %>%
  t() %>%
  reshape::melt() %>%
  arrange(X1) %>%
  rename(sample = X2) %>%
  mutate(group = ifelse(grepl("norm", sample), "normal tissue", ifelse(grepl("ct", sample), "cancer tissue", "cancer cell line"))) %>%
  ggplot(aes(x = sample, y = value, fill = group)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Sample", y = "log10(FPKM)", fill = "Group") +
  theme_bw()
```

#### Diagram on filtering criteria for Integrated Pathway Analysis

```{r gene filtering}
grViz(diagram = "digraph flowchart {
      # define node aesthetics
      node [fontname = Arial, shape = rect, color = DeepSkyBlue, style = filled, fontcolor = White]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
# set up node layout
      tab1 -> tab2;
      tab2 -> tab3;
      tab3 -> tab4;
      tab3 -> tab5
      }
      [1]: 'Initial Number of Genes: 62,635'
      [2]: 'Filter p-value < 0.001 and log2 fold change < -5 or > 5'
      [3]: '280 Genes Remaining'
      [4]: '198 Genes Downregulated'
      [5]: '82 Genes Upregulated'
      ")
```

Perform filtering of genes according to the above diagram

```{r read in expression differences file and filter}

exp <- data.table::fread("./data/gene_exp.diff", sep = "\t")

# the cutoff will be p-val of < 0.001 and log2 fold change greater than 5 or less than -5

exp_filt <- exp %>%
  arrange(`log2(fold_change)`) %>%
  filter(p_value < 0.001) %>%
  filter(`log2(fold_change)` != -Inf, # Inf or -Inf means gene had 0 expression in one of the groups
         `log2(fold_change)` != Inf) %>%
  filter(`log2(fold_change)` <= -5 | `log2(fold_change)` >= 5)
```

Separate up and down regulated genes

```{r separate upregulated and downregulated genes}
up_genes <- exp_filt %>%
  filter(`log2(fold_change)` > 0) %>%
  select(gene_id, gene)

down_genes <- exp_filt %>%
  filter(`log2(fold_change)` < 0) %>%
  select(gene_id, gene)
```



#### PCA with filtered genes only

```{r redo PCA with filtered genes only}
fpkm_filt <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm_filt)[26:29] <- names(fpkm_filt)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm_filt <- fpkm_filt %>% 
  select(tracking_id, ends_with("_FPKM")) %>%
  filter(tracking_id %in% exp_filt$test_id) %>%
  column_to_rownames("tracking_id")


fpkm_filt_norm <- log10(fpkm_filt + 0.1)          

fpkm_filt_norm <- t(fpkm_filt_norm)

fpkm_filt_norm <- fpkm_filt_norm %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  mutate(type = ifelse(grepl("norm", sample), "normal tissue",
                         ifelse(grepl("ct", sample), "cancer tissue", "cancer cell line")))
  

fpkm_filt_norm.data <- fpkm_filt_norm %>% select(-c(sample, type))

pca2 <- prcomp(fpkm_filt_norm.data)

pca2$x

autoplot(pca2, scale = FALSE, data = fpkm_filt_norm, colour = "type") + ggtitle("PCA on Filtered Genes") + theme_bw()
```

Compute Dendrograms

Use pearson correlation for distance and ward.D2 and complete for clustering method

```{r Dendrogram of differentially expressed genes ward method, warning = FALSE}
rownames(fpkm_filt_norm.data) <- fpkm_filt_norm$sample

# try pearson correlation and euclidean distance for distance measures
res.pearson.dist <- get_dist(fpkm_filt_norm.data, method = "pearson")
res.euclidean.dist <- get_dist(fpkm_norm.data, method = "euclidean")

# try ward and complete hierarchical clustering methods
res.eucl.ward.hc <- hclust(d = res.euclidean.dist, method = "ward.D2")
res.eucl.complete.hc <- hclust(d = res.euclidean.dist, method = "complete")
res.pear.ward.hc <- hclust(d = res.pearson.dist, method = "ward.D2")
res.pear.complete.hc <- hclust(d = res.pearson.dist, method = "complete")


p1 <- fviz_dend(res.pear.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "pearson", "Method:", "ward.D2", sep = " "))


p2 <- fviz_dend(res.pear.complete.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "pearson", "Method:", "complete", sep = " "))

p3 <- fviz_dend(res.eucl.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "euclidean", "Method:", "ward.D2", sep = " "))

p4 <- fviz_dend(res.eucl.complete.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE,
          main = paste("Distance:", "euclidean", "Method:", "complete", sep = " "))

grid.arrange(p1, p2, p3, p4)
```

It appears that using euclidean distance along with Ward's method gives us the clustering that we expect to see given our knowledge of the three groups.

```{r visualize p3}
fviz_dend(res.eucl.ward.hc, cex = 0.5, k = 3, color_labels_by_k = TRUE, rect = TRUE, rect_fill = TRUE, show_labels = TRUE,
          main = "Hierarchical Clustering of Samples Based on Filtered Genes", xlab = "Samples")
```




Evaluate Dendrograms

```{r p-values for ward clustering}
# create cluster for parallel computing
n.cores <- detectCores()*3/4
cl <- makePSOCKcluster(n.cores)
clusterSetRNGStream(cl, 1234)

# calculate p-value dendrogram for pearson correlation and ward.d2 method
res.pear.ward.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "ward.D2", method.dist = "correlation", nboot = 1000)
# calculate p-value dendrogram for pearson correlation and complete method
res.pear.complete.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "complete", method.dist = "correlation", nboot = 1000)
# calculate p-value dendrogram for euclidean dist and ward.d2 method
res.eucl.complete.pv <- parPvclust(cl, t(fpkm_filt_norm.data), method.hclust = "ward.D2", method.dist = "euclidean", nboot = 1000)

stopCluster(cl)

# plot
p1 <- plot(res.pear.ward.pv, hang = -1, cex = 0.5)
p2 <- plot(res.pear.complete.pv, hang = -1, cex = 0.5)
p3 <- plot(res.eucl.complete.pv, hang = -1, cex = 0.5)

# Clusters with AU >= 95% are considered to be strongly supported by data.
```

Above, it appears that using pearson correlation and complete clustering method clustered ccl3_FPKM poorly. The AU values for using both pearson and euclidean along with Ward's method gave stronger AU values.


```{r view tanglegram of top clusters}
dend1 <- as.dendrogram(res.pear.ward.hc)
dend2 <- as.dendrogram(res.eucl.ward.hc)

dendlist(dend1, dend2) %>%
  untangle(method = "step1side") %>%
  tanglegram(main_left = "Pearson Correlation", main_right = "Euclidean Distance", 
             main = "Ward.D2", lwd = 1, cex_main = 1, margin_inner = 5)
```

Using pearson correlation for distance metric instead of euclidean results in cancer cell line 2 clustering with the normal tissues instead of with the other cancer cell line samples. I have determined that using euclidean distance and ward's method to be the best choices with my data.


#### Plot Heatmap

```{r heatmap wards}
Heatmap(data.matrix(fpkm_filt_norm.data), 
        name = "Expression",
        show_column_names = FALSE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")
```


```{r annotated heatmap wards}


boxplot <- rowAnnotation(`log10(fpkm)` = anno_boxplot(data.matrix(fpkm_filt_norm.data)),
                         gp=gpar(fill="#CCCCCC"))

hmap <- Heatmap(data.matrix(fpkm_filt_norm.data), 
        name = "Expression",
        show_column_names = FALSE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2",
        right_annotation = boxplot)

draw(hmap)
```


At this point, I submitted the list of up regulated and down regulated genes to do a simple integrated pathway analysis (software from QIAGEN). There were many significant pathways returned, many of the down regulated ones were involved in immune system function whereas many of the up regulated pathways have been previously implicated in cancer. The excel files are located in the IPA-results folder.

Moving forward, I will be looking at individual genes that are involved in a specific pathway of interest to me, carbohydrate metabolism, which was down regulated in cancerous samples, compared to normal samples.



#### Heatmap of carbohydrate metabolism pathway.

I will look at the "metabolism of carbohydrates" pathway, which is downregulated in gi cancer.

```{r get list of genes in metabolism of carbohydrates pathway}
meta <- read_xls("./IPA-results/down-genes_diseases_and_functions.xls", col_names = TRUE) %>%
  filter(`Diseases or Functions Annotation` == "Metabolism of carbohydrate") %>%
  unnest_tokens(genes, Molecules) %>%
  mutate(genes = toupper(genes)) %>%
  select(genes)

meta
```

```{r create df containing FPKM values for genes in carbohydrate metabolism pathway}
fpkm <- data.table::fread("./data/genes.fpkm_tracking", sep = "\t")

names(fpkm)[26:29] <- names(fpkm)[26:29] %>% # fix column names
  str_replace("ct2", "ct3")


fpkm <- fpkm %>% 
  select(gene_short_name, ends_with("_FPKM"))

meta <- merge(x = meta,
      y = fpkm, 
      all.x = TRUE,
      all.y = FALSE,
      by.x = "genes",
      by.y = "gene_short_name")


meta <- meta %>% 
  column_to_rownames("genes") %>% 
  t()


meta_norm <- log10(meta + 0.1)
```

```{r simple heatmap of normalized metabolism pathway fpkms}
Heatmap(meta_norm, 
        name = "expression",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")
```

```{r annotated heatmap, warning=FALSE}
Heatmap(meta_norm, 
        name = "expression",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2") + 
  rowAnnotation(`log10(fpkm)` = anno_boxplot(meta_norm)) +
  rowAnnotation(rn = anno_text(rownames(meta_norm)))
```



Using the same clustering method (euclidean distance and ward.D2 method) it appears that cancer cell line 2 becomes its own cluster, and cancer cell lines 1 and 3 cluster with the cancer tissue.

```{r other clustering method testing}
# compare other clustering methods with original to check to see if it is still the best method for the selected pathway

# original clustering method
ht1 <- Heatmap(meta_norm, 
        name = "ht1",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2")


# pearson correlation with ward.D2 clustering
ht2 <- Heatmap(meta_norm, 
        name = "ht2",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2",
        column_names_gp = gpar(fontsize = 9))

# euclidean distance with complete clustering 

ht3 <- Heatmap(meta_norm, 
        name = "ht3",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "euclidean",
        clustering_distance_columns = "euclidean",
        clustering_method_rows = "complete",
        clustering_method_columns = "complete",
        column_names_gp = gpar(fontsize = 9))

# pearson correlation with ward.D2 clustering

ht4 <- Heatmap(meta_norm, 
        name = "ht4",
        show_column_names = TRUE, 
        row_split = 3,
        clustering_distance_rows = "pearson",
        clustering_distance_columns = "pearson",
        clustering_method_rows = "complete",
        clustering_method_columns = "complete",
        column_names_gp = gpar(fontsize = 9))

ht1 + ht2 + ht3 + ht4
```

When plotting all 4 heatmaps together with the original heatmap as the main, the separation is clearest with the original. Individually, ht2 makes cell line 3 its own cluster and clusters cell line 2 and cancer tissue 3 with normal, ht2 is similar to original heatmap, and ht3 makes cell line 1 and 3 are their own clusters. I conclude that the original clustering method is still the best option for this pathway.

#### Linear model test on individual genes for significance

```{r add principal components to data frame}

# add PC1 and PC2 to the data frame for correction

lm_df <- merge(x = pca$x, y = meta_norm, by = "row.names") %>%
  select(-c(4:9)) %>%
  rename(group = Row.names) %>%
  mutate(group = if_else(grepl("norm", group), "normal tissue", if_else(grepl("ct", group), "cancer tissue", "cancer cell line")))

lm_df
```

```{r lm test}
# function to return p-value from lm object
lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}


# lm test on all genes
no_correction <- list() # initialize empty list to hold p-values for non-corrected lm test
correction <- list() # empty list to hold p-values for corrected lm test
for (i in names(lm_df[4:length(names(lm_df))])){
  no_correction[i] <- lmp(lm(get(i) ~ group, lm_df))
  correction[i] <- lmp(lm(get(i) ~ group + PC1 + PC2, lm_df))
}


# store results in data frame
meta_pval <- rbind(as.data.frame(no_correction), as.data.frame(correction)) %>% 
  t() %>%
  as.data.frame() %>%
  rename("Non PC Corrected p-value" = V1, "PC Corrected p-value" = V2) %>%
  format(scientific = FALSE) %>%
  rownames_to_column("gene")

meta_pval <- meta_pval %>%
  mutate(`Non PC Corrected p-value` = as.numeric(`Non PC Corrected p-value`),
         `PC Corrected p-value` = as.numeric(`PC Corrected p-value`)) 

```


Plot p-values for each gene's significance in determining difference between cancer and non cancer

```{r plot p-values, message=FALSE}
meta_pval %>%
  reshape::melt() %>%
  #mutate(value = as.numeric(value)) %>%
  arrange(gene) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(gene), value), y = value)) +
  geom_bar(aes(fill = variable), stat = "identity", position = "dodge", width = 0.5) +
  scale_fill_manual(values = c("#1B32FA", "#FA871B")) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  geom_hline(yintercept = 0.05, linetype = "dashed") +
  labs(y = "p-value", x = "gene")
```

Plot heatmap of carbohydrate gene expression

```{r heatmap of carbohydrate metabolism pathway}
Heatmap(meta_norm,
        name = "expression",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2") + 
  rowAnnotation(`log10(fpkm)` = anno_boxplot(meta_norm)) +
  rowAnnotation(rn = anno_text(rownames(meta_norm)))
```



It appears that after correcting for the first two principal components, all genes except for CD22, B3GNT8, B3GAT1, and ALDH1A1 are significant.

* P2RY6: [promotes tumorigenesis by inhibiting apoptosis](https://pubmed.ncbi.nlm.nih.gov/29454075/)

* PIK3CG: [inhibits PI3K-Akt/PKB signaling system responsible for tumorigenesis and the progression of colorectal cancers](https://pubmed.ncbi.nlm.nih.gov/12473596/)

* **GPD1**: [encodes glycerol-3-phosphate dehydrogenase, which plays a crucial role in carbohydrate and lipid metabolism by catalyzing conversion of DHAP and NADG to G3P (glycerol-3-phosphate) and NAD+.](https://en.wikipedia.org/wiki/Glycerol-3-phosphate_dehydrogenase_1) Has been implicated as a tumor suppressor in [breast cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5731876/) and [bladder cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9284842/)

* CSFR1: [involved in production, differentiation, and function of macrophages](https://en.wikipedia.org/wiki/Colony_stimulating_factor_1_receptor)

* **GCNT4**: [predicted to be involved in O-glycan processing and carbohydrate metabolic process.](https://www.ncbi.nlm.nih.gov/gene/51301). [Found to be downregulated in gastric cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7457769/)

* PRKCB: [encodes protein involved in B cell activation, apoptosis induction, endothelial cell proliferation, and intestinal sugar absorption.](https://www.ncbi.nlm.nih.gov/gene/5579) [Previously implicated in colorectal cancer](https://link.springer.com/article/10.1007/s12253-013-9739-5)

* B4GALNT2: [catalyzes last step in biosynthesis of human Sd(a) antigen and Cad antigen.](https://www.genecards.org/cgi-bin/carddisp.pl?gene=B4GALNT2) [Found to be highly expressed in model triple negative breast cancer](https://pubmed.ncbi.nlm.nih.gov/34589428/)

* CEBPA: [encodes transcription factor involved in maturation of certain blood cells and believed to act as a tumor supressor.](https://medlineplus.gov/genetics/gene/cebpa/) [Implicated in acute myeloid leukemia](https://www.frontiersin.org/articles/10.3389/fonc.2022.806137/full)

* SST: [regulator of endocrine system through interactions with pituitary growth hormone, thyroid stimulating hormone, and most hormones of GI tract. Also affects rates of neurotransmission in the CNS and proliferation of normal and tumorigenic cells](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SST)

* **SLC2A4**: [encodes member of K+ dependent Na/Ca exchanger protein family](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SLC24A4) Not many articles on its implications in cancer.

* PTPRC: [known to be signaling molecules that regulate a variety of cellular processes including cell growth, differentiation, mitosis, and oncogenic transformation](https://www.ncbi.nlm.nih.gov/gene/5788)

* GCNT2: [encodes enzyme responsible for formation of blood group I antigen.](https://en.wikipedia.org/wiki/GCNT2) [Hypomethylated in colorectal cancer](https://pubmed.ncbi.nlm.nih.gov/25750292/)


I have selected three genes of interest to me: GPD1, GCNT4, and SLC24A for further analysis.

```{r}
summary(lm(GPD1 ~ group, lm_df))
```

```{r barplots of individual genes, fig.height=12, fig.width=12}
p1 <- lm_df %>%
  select(group, GPD1) %>%
  group_by(group) %>%
  summarise(mean = mean(GPD1), sd = sd(GPD1)) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(group), mean), y = mean, fill = group)) +
  geom_col() +
  geom_errorbar(aes(x = group, ymin = mean - sd, ymax = mean + sd), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_text(aes(label = paste("Average log10(fpkm):", round(mean, 2), sep = " "),
                hjust = 0.5,
                vjust = ifelse(mean < 0, 5.5, -2.5)),
            size = 3) +
  geom_hline(yintercept = 0) +
  ylim(-2, 2) +
  labs(x = "", y = "Average log(10) FPKM", title = "GPD1") +
  annotate("text", label = paste("p-value:", round(lmp(lm(GPD1 ~ group, lm_df)), 6), sep = " "), x = 2, y = 2)

p2 <- lm_df %>%
  select(group, SLC2A4) %>%
  group_by(group) %>%
  summarise(mean = mean(SLC2A4), sd = sd(SLC2A4)) %>%
  ggplot(aes(x = forcats::fct_reorder(factor(group), mean), y = mean, fill = group)) +
  geom_col() +
  geom_errorbar(aes(x = group, ymin = mean - sd, ymax = mean + sd), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_text(aes(label = paste("Average log10(fpkm):", round(mean, 2), sep = " "),
                hjust = 0.5,
                vjust = ifelse(mean < 0, 5.5, -2.5)),
            size = 3) +
  geom_hline(yintercept = 0) +
  ylim(-2, 2) +
  labs(x = "", y = "Average log(10) FPKM", title = "SLC2A4") +
  annotate("text", label = paste("p-value:", round(lmp(lm(SLC2A4 ~ group, lm_df)), 6), sep = " "), x = 2, y = 2)


grid.arrange(p1, p2, ncol = 1)
```








```{r annotated heatmap with p-value annotations, warning=FALSE, include=FALSE, eval=FALSE}
# add p-values as column annotations

h1 <- Heatmap(meta_norm,
        name = "expression",
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2") + 
  rowAnnotation(`log10(fpkm)` = anno_boxplot(meta_norm)) +
  rowAnnotation(rn = anno_text(rownames(meta_norm)))

pvals <- list()
for (i in column_order(h1)){
  pvals <- c(pvals, meta_pval[i, 3])
}

pvals <- unlist(pvals)
is_sig <- pvals < 0.05
pch <- rep("*", length(pvals))
pch[!is_sig] <- NA

# color mapping for p-values
pvalue_col_fun <- colorRamp2(c(0, 2, 3), c("yellow", "white", "red")) 


ha <- HeatmapAnnotation(
    pvalue = anno_simple(pvals, col = pvalue_col_fun, pch = pch),
    annotation_name_side = "left")

ht2 <- Heatmap(meta_norm,
        name = "expression",
        top_annotation = ha,
        show_column_names = TRUE, 
        row_split = 3, 
        clustering_method_rows = "ward.D2",
        clustering_method_columns = "ward.D2") + 
  rowAnnotation(`log10(fpkm)` = anno_boxplot(meta_norm)) +
  rowAnnotation(rn = anno_text(rownames(meta_norm)))

# define legend for p-value
lgd_pvalue <- Legend(title = "p-value", col_fun = pvalue_col_fun, at = c(0, 1, 2, 3, 4), 
    labels = c("0.5", "0.05", "0.01", "0.001", "0.0001"))

# legend for significant p-values
lgd_sig <- Legend(pch = "*", type = "points", labels = "< 0.05")

draw(ht2, annotation_legend_list = list(lgd_pvalue, lgd_sig))
```
